SNIPS/snip-20.md

```diff
@@ -0,0 +1,120 @@
+
+A _paymaster transaction_ is one with a non-empty _paymaster field_. A valid paymaster transaction specifies three concatenated pieces of data in the `paymaster` field:
+1. Paymaster contract address
+2. ERC-20 token contract address
+3. Maximal allowed exchange rate between the above token and STRK. Specifically, max rate $r$ means the user is willing to pay at most $r$ tokens for a single STRK.
+
+We now outline the steps for processing of a paymaster transaction by the sequencer. For simplicity, the initial flow will involve charging at the user's max rate $r$; the optimization of charging at discounted rates will be deferred to the next section.
+
+#### Mempool
+
+Users can submit unrealistically low rates; these should be interpreted as overvaluations of their designated payment tokens. Exaggerated claims should be rejected. However, we wish to avoid defining "reasonable rates" at the protocol level. To this end, each sequencer client locally defines reasonable rates via dictionary: `TOKEN → min_STRK/TOKEN_rate`. (The ratio follows exchange conventions: it is the price of one STRK in units of the TOKEN.) This dictionary can be static or dynamically updated e.g by some oracle feed. Sequencers can also choose to serve only some chosen whitelist of paymaster contracts.
+
+We propose that full nodes indiscriminantly propagate paymaster transactions without looking into the `paymaster` field; only sequencers will bar entry into their respective mempools.
+
+Now the flow:
+
+1. Syntactic checks (transaction is not a cat gif).
+2. User submitted rate should be at least client's locally configured rate. $$\mathtt{min\_ STRK/TOKEN\_ rate}\leq r$$
+3. Balance checks:
+    * User token balance should cover their bid given their specified max exchange rate $$\mathtt{balance(user)}\geq \mathtt{max\_ amount}\times r(\mathtt{base\_ price(current\_ block)}+\mathtt{tip}).$$
+    * Paymaster STRK balance should cover user bid $$\mathtt{balance(paymaster)}\geq \mathtt{max\_ amount}\times (\mathtt{base\_ price(current\_ block)}+\mathtt{tip}).$$
+
+#### Block builder
+
+A paymaster contract should be used only by those who trust its code. Such users know the paymaster's exchange policies, and therefore know which rates they should submit to pass, and which rates are too frugal and may cause `__validate_paymaster__` to fail. Hence, we consider it justified to "blame" users for failed `__validate_paymaster__` and to consequently charge them. Now arises a problematic scenario: the user may hold zero STRK and the paymaster rejected the request to cover its transaction fee. In this particular case we allow the sequencer to charge the transaction fee directly in the user's designated token.
```

Comment:
I would expect the exchange rate of the sequencers to be higher than the exchange rate of the paymasters.

Reply:
Sequencers will set a low threshold just to filter overvaluations. If they set their rates too high, they will effectively block out paymaster transactions. What are you concerned about?

Reply:
If the sequencer gives me a better conversion rate than the paymasters.
I would move my logic to validate, use a paymaster that always revets, and pay the sequencer with my desired token.

Reply:
"Sequencers will set a low threshold just to filter overvaluations." does not seem like a rational behavior.

I think Sequencers will set high threshold and actually paymaster will use r/2 or something like that.
maybe we want to do that at the protocol layer?